### PR TITLE
[SYCL][L0] Move command list cache usage under mutex

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**
For the example of usage that may cause issue please refer to assert_in_simulteneous_kernels E2E test. n queues that are working on the same device and on default context will work with the same zeCompute/CopyCommandListCache without protection since mutexes for queues are different while cache is the same.
Operator[] may trigger insertion and rehash so simulteneous access is not thread-safe. In this code cache is not prefilled and insertion is done on the first access for every unique ZeDevice what may end up with simultaneous write access and memory corruption

Signed-off-by: Tikhomirova, Kseniya <kseniya.tikhomirova@intel.com>

### CODE REVIEW

**Comment by KseniyaTikhomirova** [comment]
OCL failure - known failure, expecting https://github.com/intel/llvm-test-suite/pull/944
ESIMD failure is also not possible to be related since I did a small change in L0 plugin.

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1108,19 +1108,20 @@ _pi_context::getAvailableCommandList(pi_queue Queue,
...
-
   // Initally, we need to check if a command list has already been created
   // on this device that is available for use. If so, then reuse that
   // Level-Zero Command List and Fence for this PI call.
   {
     // Make sure to acquire the lock before checking the size, or there
     // will be a race condition.
     std::lock_guard<std::mutex> lock(Queue->Context->ZeCommandListCacheMutex);
+    // Under mutex since operator[] does insertion on the first usage for every
+    // unique ZeDevice.
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1119]: > The first time we try to use command list cache for exact device, operator[] does insertion and use this command list cache in the following code

That's exactly that I think is wrong.
Instead we need the `if (ZeCommandListCache.size() > 0) {` to return false, and new true item is inserted to the cache (not default-initialized one).

[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1119]: but we still do not insert default initialized items into the command cache list for exact device.
Here we are creating the list for exact device, not inserting the items into it.

This operator[] does insertion to map for exact ZeDevice, the list is still empty.
  std::unordered_map<ze_device_handle_t, std::list<ze_command_list_handle_t>>
      ZeComputeCommandListCache;

**Pull request merged**